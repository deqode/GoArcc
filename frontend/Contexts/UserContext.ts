import { createContext } from "react";

export const UserContext = createContext({
    user: {
        accessToken: "",
        idToken: "",
        userId: "",
        provider: "",
        state: 0,
        //  loading=0
        //  authenticated=1
        //  unauthenticated=-1
    },
    setUser: (value) => { },
    removeUser: () => { }
});
