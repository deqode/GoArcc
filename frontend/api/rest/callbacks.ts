import axios from 'axios'
import { ResponseError, UserResponse } from '../../interface'

import { SERVER } from '../../utils/constants'

export const getAuth0Callback = async (code: string, state: string): Promise<UserAuthResponse> => {
  if (code !== '' && state !== '')
    try {
      const response = await axios.get(
        `${SERVER}/authentication/callback?code=${code}&state=${state}`
      )
      if (response.data && response.data.userId) {
        return {
          userId: response.data.userId,
          idToken: response.data.idToken,
          accessToken: response.data.idToken,
          error: false,
          message: '',
        }
      } else
        return {
          userId: '',
          idToken: '',
          accessToken: '',
          error: true,
          message: 'Data Not found',
          // todo : integrate with error message from backend
        }
    } catch (e) {
      return {
        userId: '',
        idToken: '',
        accessToken: '',
        error: true,
        message: 'Network Error',
        // todo : integrate with error message from backend
      }
    }
  else
    return {
      userId: '',
      idToken: '',
      accessToken: '',
      error: true,
      message: 'Invalid types',
      // todo : integrate with error message from backend
    }
}

export const getVCSConnectionGitHubCallback = async (
  code: string,
  accountId: string,
  idToken: string
): Promise<ResponseError> => {
  if (code !== '' && accountId !== '')
    try {
      const response = await axios.get(
        `${SERVER}/vcs-connection/GITHUB/callback?code=${code}&account_id=${accountId}`,
        { headers: { Authorization: `Bearer ${idToken}` } }
      )
      if (response.data && response.data.accountId) {
        return {
          error: false,
          message: '',
        }
      } else
        return {
          error: true,
          message: 'Data Not found',
          // todo : integrate with error message from backend
        }
    } catch (e) {
      return {
        error: true,
        message: 'Network Error',
        // todo : integrate with error message from backend
      }
    }
  else
    return {
      error: true,
      message: 'Invalid types',
      // todo : integrate with error message from backend
    }
}

interface UserAuthResponse extends ResponseError, UserResponse {}
