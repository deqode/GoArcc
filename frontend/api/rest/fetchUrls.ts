import axios from 'axios'
import { ResponseError } from '../../interface'

import { SERVER } from '../../utils/constants'

export const getLoginURL = async (): Promise<LoginURL> => {
  try {
    const response = await axios.get(`${SERVER}/authentication/login`)
    if (response.data && response.data.url) {
      return {
        url: response.data.url,
        error: false,
        message: '',
      }
    } else
      return {
        url: '',
        error: true,
        message: 'Data Not found',
        // TODO : integrate with error message from backend
      }
  } catch (e) {
    return {
      url: '',
      error: true,
      message: 'Network Error',
      // TODO : integrate with error message from backend
    }
  }
}

export const getGithubVCSConnection = async (idToken: string): Promise<GithubVCSConnection> => {
  try {
    const response = await axios.get(`${SERVER}/vcs-connection/authorize/GITHUB`, {
      headers: { Authorization: `Bearer ${idToken}` },
    })
    if (response.data && response.data.redirectUrl) {
      return {
        redirectUrl: response.data.redirectUrl,
        error: false,
        message: '',
      }
    } else
      return {
        redirectUrl: '',
        error: true,
        message: 'Data Not found',
        // TODO : integrate with error message from backend
      }
  } catch (e) {
    return {
      redirectUrl: '',
      error: true,
      message: 'Network Error',
      // TODO : integrate with error message from backend
    }
  }
}

interface GithubVCSConnection extends ResponseError {
  redirectUrl: string
}
interface LoginURL extends ResponseError {
  url: string
}
