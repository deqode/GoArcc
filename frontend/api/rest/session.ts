import axios from 'axios'

import { ResponseError, UserResponse } from '../../interface'

export const setUserSession = async (data: UserResponse): Promise<ResponseError> => {
  // TODO:handle data
  try {
    const response = await axios.post(`/api/session/set`, data, {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    if (response.data.success) {
      return {
        error: false,
        message: '',
      }
    } else
      return {
        error: true,
        message: 'Server Error',
        // TODO : integrate with error message from backend
      }
  } catch (e) {
    return {
      error: true,
      message: 'Network Error',
      // TODO : integrate with error message from backend
    }
  }
}

export const destroyUserSession = async (): Promise<ResponseError> => {
  try {
    const response = await axios.post(
      `/api/session/destroy`,
      {},
      {
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      }
    )
    if (response.data.success) {
      return {
        error: false,
        message: '',
      }
    } else
      return {
        error: true,
        message: 'Server Error',
        //  : integrate with error message from backend
      }
  } catch (e) {
    return {
      error: true,
      message: 'Network Error',
      // TODO : integrate with error message from backend
    }
  }
}
