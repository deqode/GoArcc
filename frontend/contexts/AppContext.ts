// import Centrifuge from 'centrifuge'
// import { createContext } from 'react'
// import { AppContextInterface } from '../interface'
// import { CENTRIFUGO } from '../utils/constants'

// const appContext: AppContextInterface = {
//   app: {
//     centrifuge: new Centrifuge(CENTRIFUGO),
//     subscribed: false,
//   },
//   setApp: () => null,
//   removeApp: () => null,
// }

// export const AppContext = createContext<AppContextInterface>(appContext)
export {}
