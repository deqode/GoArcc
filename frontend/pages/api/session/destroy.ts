import { withIronSession } from 'next-iron-session'
import { sessionCongfig } from '../../../utils/constants'
import { withSentry } from '@sentry/nextjs'

// TODO: notify platform to invalidate
// TODO:validate
const handler = async (req: any, res: any) => {
  req.session.destroy('user')
  res.json({ susscess: true })
}

export default withSentry(withIronSession(handler, sessionCongfig))
