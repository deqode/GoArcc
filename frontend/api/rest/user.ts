import { ResponseError, UserAccount } from '../../intefaces/interface'
import { get } from '../../services/rest'
import { SERVER } from '../../utils/constants'

export interface AllUserAccounts extends ResponseError<AllUserAccounts> {
  accounts: Array<UserAccount>
}

export const getAllUserAccounts = async (
  userId: string,
  idToken: string
): Promise<AllUserAccounts> => {
  // TODO:valide inputs
  const response = await get<AllUserAccounts>(`${SERVER}/account/get-user-accounts/${userId}`, {
    Authorization: `Bearer ${idToken}`,
  })
  if (response.error) {
    return { accounts: [], ...response }
  }
  if (response.data && response.data.accounts) {
    return {
      error: false,
      message: '',
      accounts: response.data.accounts,
    }
  } else
    return {
      error: true,
      message: 'Data Not found',
      accounts: [],
    }
}
