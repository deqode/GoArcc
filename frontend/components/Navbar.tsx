import { AppBar, Button, Grid, Toolbar } from '@material-ui/core'
import { makeStyles } from '@material-ui/styles'

import Link from 'next/link'
import { useRouter } from 'next/router'
import Image from 'next/image'

import React, { ReactElement, useContext } from 'react'
import { destroyUserSession } from '../api/rest/session'
import UserContext from '../contexts/UserContext'

const useStyles = makeStyles({
  root: {
    width: '60px',
  },
})

function Navbar(): ReactElement {
  const { user } = useContext(UserContext)
  const router = useRouter()
  const classes = useStyles()
  return (
    <AppBar position="static" color="transparent">
      <Toolbar>
        <Grid justify="space-between" container>
          <Grid item className={classes.root}>
            <Link href="/">
              <Image src="/assets/logo.png" width={500} height={500} alt="logo" />
            </Link>
          </Grid>
          <Grid item>
            {user.loggedIn ? (
              <Button
                onClick={async () => {
                  const res = await destroyUserSession()
                  if (!res.error) {
                    router.reload()
                  } else {
                    router.push('/error', { query: { message: 'Network Error' } })
                  }
                }}
                color="inherit">
                Logout
              </Button>
            ) : (
              ''
            )}
          </Grid>
        </Grid>
      </Toolbar>
    </AppBar>
  )
}

export default Navbar
