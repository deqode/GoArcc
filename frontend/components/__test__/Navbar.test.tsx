import { render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'

import * as api from '../../api/rest/session'
import { User } from '../../contexts/UserContext'
import Navbar from '../Navbar'

it('should match the snapshot of logged-in user', () => {
  const mockUser: User = {
    idToken: 'mockidToken',
    loggedIn: true,
  }
  const tree = renderer.create(<Navbar user={mockUser} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should match the snapshot of logged-out user', () => {
  const mockUser: User = {
    idToken: 'mockidToken',
    loggedIn: false,
  }
  const tree = renderer.create(<Navbar user={mockUser} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should click logout ', async () => {
  const mockUser: User = {
    idToken: 'mockidToken',
    loggedIn: true,
  }
  const { getByText } = render(<Navbar user={mockUser} />)
  fireEvent.click(getByText('common:LOGOUT'))
})

it('should get logged out form session  ', async () => {
  api.destroyUserSession = jest.fn().mockReturnValue(() => Promise.resolve({ error: false }))
  const mockUser: User = {
    idToken: 'mockidToken',
    loggedIn: true,
  }
  const { getByText } = render(<Navbar user={mockUser} />)
  fireEvent.click(getByText('common:LOGOUT'))
})
