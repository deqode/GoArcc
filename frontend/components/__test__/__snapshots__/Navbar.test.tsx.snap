// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot of logged-in user 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorTransparent MuiPaper-elevation0"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-between"
    >
      <div
        className="MuiGrid-root makeStyles-logo-1 MuiGrid-item"
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "margin": 0,
              "maxWidth": "100%",
              "overflow": "hidden",
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "display": "block",
                "maxWidth": "100%",
              }
            }
          >
            <img
              alt=""
              aria-hidden={true}
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjMwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style={
                Object {
                  "border": "none",
                  "display": "block",
                  "margin": 0,
                  "maxWidth": "100%",
                  "padding": 0,
                }
              }
            />
          </div>
          <noscript>
            <img
              alt="logo"
              decoding="async"
              onClick={[Function]}
              onMouseEnter={[Function]}
              src="/assets/logo.png"
              srcSet="/_next/image?url=%2Fassets%2Flogo.png&w=640&q=75 1x, /_next/image?url=%2Fassets%2Flogo.png&w=1080&q=75 2x"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": undefined,
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </noscript>
          <img
            alt="logo"
            decoding="async"
            onClick={[Function]}
            onMouseEnter={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            common:LOGOUT
          </span>
        </button>
      </div>
    </div>
  </div>
</header>
`;

exports[`should match the snapshot of logged-out user 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorTransparent MuiPaper-elevation0"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-between"
    >
      <div
        className="MuiGrid-root makeStyles-logo-1 MuiGrid-item"
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "margin": 0,
              "maxWidth": "100%",
              "overflow": "hidden",
              "position": "relative",
            }
          }
        >
          <div
            style={
              Object {
                "boxSizing": "border-box",
                "display": "block",
                "maxWidth": "100%",
              }
            }
          >
            <img
              alt=""
              aria-hidden={true}
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjMwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style={
                Object {
                  "border": "none",
                  "display": "block",
                  "margin": 0,
                  "maxWidth": "100%",
                  "padding": 0,
                }
              }
            />
          </div>
          <noscript>
            <img
              alt="logo"
              decoding="async"
              onClick={[Function]}
              onMouseEnter={[Function]}
              src="/assets/logo.png"
              srcSet="/_next/image?url=%2Fassets%2Flogo.png&w=640&q=75 1x, /_next/image?url=%2Fassets%2Flogo.png&w=1080&q=75 2x"
              style={
                Object {
                  "border": "none",
                  "bottom": 0,
                  "boxSizing": "border-box",
                  "display": "block",
                  "height": 0,
                  "left": 0,
                  "margin": "auto",
                  "maxHeight": "100%",
                  "maxWidth": "100%",
                  "minHeight": "100%",
                  "minWidth": "100%",
                  "objectFit": undefined,
                  "objectPosition": undefined,
                  "padding": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 0,
                }
              }
            />
          </noscript>
          <img
            alt="logo"
            decoding="async"
            onClick={[Function]}
            onMouseEnter={[Function]}
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      />
    </div>
  </div>
</header>
`;
